### DISCUSSION TITLE
keycloak-js and Forward Query Parameters

### DISCUSSION THREAD

**Discussion by Jorge1701** [keycloak/keycloak#13094] [Ideas]
Hi, I'm wondering if there is a way to pass forward parameters using the js adapter. As far as I know the is no way of doing so, or I'm misunderstanding the way these parameters work.

# Context
We have a client that wants to use their own IDP to authenticate their users, for that we just see the domain they are requesting from and use the `idpHint` property in `KeycloakLoginOptions` when we call the login function to fill the `kc_idp_hint` query param and it goes to the correct IDP. This works as expected.

# Problem
Now our client wants to add an `api_key` parameter to the url that it redirects to their IDP, if that key is not there or is incorrect we are not authorized to access their IDP.

For that we configure this by adding the `api_key` in the Forwarded Query Parameters in the OpenID Connect Config of that provider like this:

![imagen](https://user-images.githubusercontent.com/26526255/178799087-da053c68-15c6-49c1-a0c1-5f47ba99ea98.png)

As I understand it, these are the parameters that if we send along side the `kc_idp_hint` they will be forwarded to our clients IDP.

The problem is that I cant see a way of sending these custom parameters in the login function in the js adapter.

# Solution
I have found that I can make the following change to the `createLoginUrl` function in `adapters/oidc/js/src/keycloak.js:425`and it will support that.

```js
            if (options && options.forwardParameters) {
                url = Object.keys(options.forwardParameters).reduce(function (url, forwardParameter) {
                    return url + "&" + forwardParameter + "=" + encodeURIComponent(options.forwardParameters[forwardParameter]);
                }, url);
            }
```
After that I can just extend `KeycloackLoginOptions` in my angular 2+ app and use that with the keycloak-angular wrapper when I call the login function. Like this:

```js
import {KeycloakLoginOptions} from "keycloak-js";

export interface CustomKeycloakLoginOptions extends KeycloakLoginOptions {
  forwardParameters?: { [forwardParameter: string]: any };
}
```

Or maybe just add that `forwardParameters` property to the original `KeycloakLoginOptions` in `adapters/oidc/js/dist/keycloak.d.ts`.

# Question
Is this an intended use for the Forwarded Query Parameters and if so, is the js adapter just missing a way of passing these forwarded parameters?

If this is correct I can create a PR for that change, which I think is generic enough and does not only contemplate our `api_key` case.

**Comment by vannDS** [comment]
Adding a comment in hopes that this will be pushed up the stack. We could use this functionality for our integration because our client needs dynamic query params added into our authorization url.

**Comment by benzitohhh** [comment]
Hi there - any update on this? We also need this functionality. Thanks!
